cmd/metal-api/internal/grpc/waitServer.go

```diff
@@ -0,0 +1,138 @@
+package grpc
+
+import (
+	"fmt"
+	"github.com/metal-stack/metal-api/cmd/metal-api/internal/datastore"
+	"github.com/metal-stack/metal-api/cmd/metal-api/internal/metal"
+	v1 "github.com/metal-stack/metal-api/pkg/api/v1"
+	"github.com/metal-stack/metal-lib/bus"
+	"github.com/metal-stack/metal-lib/zapup"
+	"github.com/spf13/viper"
+	"go.uber.org/zap"
+	"math/rand"
+	"sync"
+	"time"
+)
+
+const (
+	receiverHandlerTimeout = 15 * time.Second
+	allocationTopicTTL     = time.Duration(30) * time.Second
+)
+
+func timeoutHandler(err bus.TimeoutError) error {
+	zapup.MustRootLogger().Sugar().Error("Timeout processing event", "event", err.Event())
+	return nil
+}
+
+type WaitServer struct {
+	bus.Publisher
+	ds        *datastore.RethinkStore
+	logger    *zap.SugaredLogger
+	queueLock *sync.RWMutex
+	queue     map[string]chan bool
+}
+
+func NewWaitServer(ds *datastore.RethinkStore, publisher bus.Publisher) (*WaitServer, error) {
+	tlsCfg := &bus.TLSConfig{
+		CACertFile:     viper.GetString("nsqd-ca-cert-file"),
+		ClientCertFile: viper.GetString("nsqd-client-cert-file"),
+	}
+	c, err := bus.NewConsumer(zapup.MustRootLogger(), tlsCfg, viper.GetString("nsqlookupd-http-addr"))
+	if err != nil {
+		return nil, err
+	}
+
+	s := &WaitServer{
+		Publisher: publisher,
+		ds:        ds,
+		logger:    zapup.MustRootLogger().Sugar(),
```

Comment:
I think, I would prefer passing the logger as a parameter such that the caller can decide which kind of relationship the logger has to the root logger. Also, I would like to get rid off our zapup package because it does have extra code that is memory expensive.